aten/src/ATen/templates/Functions.h

```diff
@@ -64,19 +64,87 @@ TORCH_API at::Tensor conv3d(
     const Tensor& input, const Tensor& weight, const Tensor& bias, IntArrayRef stride,
     std::initializer_list<int64_t> padding, IntArrayRef dilation = 1, int64_t groups = 1);
 
-namespace {
-  inline std::vector<int64_t> zero_sizes(const TensorOptions& options) {
-    if (options.has_memory_format()) {
-      auto memory_format = *options.memory_format_opt();
-      if (at::MemoryFormat::ChannelsLast == memory_format) {
-        return {0, 0, 0, 0};
-      }
-      if (at::MemoryFormat::ChannelsLast3d == memory_format) {
-        return {0, 0, 0, 0, 0};
-      }
-    }
-    return {0};
+namespace detail {
+
+TORCH_API void noopDelete(void*);
+
+} // namespace detail
+
+/// Provides a fluent API to construct tensors from external data.
+///
+/// The fluent API can be used instead of `from_blob` functions in case the
+/// required set of parameters does not align with the existing overloads.
+///
+///     at::Tensor tensor = at::forBlob(data, sizes)
+///             .withStrides(strides)
+///             .withContext(context, [](void *ctx) { delete static_cast<Ctx*>(ctx); })
+///             .withOptions(...)
+///             .makeTensor();
+///
+class TORCH_API TensorMaker {
+  friend TensorMaker forBlob(void* data, IntArrayRef sizes) noexcept;
+
+ public:
+  using ContextDeleter = DeleterFnPtr;
+
+  TensorMaker& withStrides(optional<IntArrayRef> value) noexcept {
+    strides_ = value;
+
+    return *this;
+  }
+
+  TensorMaker& withDeleter(std::function<void(void*)> value) noexcept {
+    deleter_ = std::move(value);
+
+    return *this;
+  }
+
+  TensorMaker& withContext(void* value, ContextDeleter deleter = nullptr) noexcept {
+    ctx_ = std::unique_ptr<void, ContextDeleter>{
+        value, deleter ? deleter : detail::noopDelete};
+
+    return *this;
+  }
+
+  TensorMaker& withTargetDevice(optional<Device> value) noexcept {
+    device_ = value;
+
+    return *this;
   }
+
+  TensorMaker& withOptions(TensorOptions value) noexcept {
+    opts_ = value;
+
+    return *this;
+  }
+
+  Tensor makeTensor();
+
+ private:
+  explicit TensorMaker(void* data, IntArrayRef sizes) noexcept
+      : data_{data}, sizes_{sizes} {}
+
+  std::size_t computeStorageSize() const noexcept;
+
+  DataPtr makeDataPtrFromDeleter() const;
+
+  DataPtr makeDataPtrFromContext() noexcept;
+
+  SmallVector<std::int64_t, 5> makeTempSizes() const noexcept;
+
+  Tensor makeEmptyTensor() const;
+
+  void* data_;
+  IntArrayRef sizes_;
+  optional<IntArrayRef> strides_{};
+  std::function<void(void*)> deleter_{};
+  std::unique_ptr<void, ContextDeleter> ctx_{nullptr, detail::noopDelete};
+  optional<Device> device_{};
+  TensorOptions opts_{};
+};
+
+inline TensorMaker forBlob(void* data, IntArrayRef sizes) noexcept {
```

Comment:
camel case function name is inconsistent with other APIs (case in point: it's `at::from_blob`)